base/sort.jl

```diff
@@ -1407,7 +1413,37 @@ julia> v
  2
 ```
 """
-sort(v::AbstractVector; kws...) = sort!(copymutable(v); kws...)
+function sort(v; kws...)
+    size = IteratorSize(v)
+    size == HasShape{0}() && throw(ArgumentError("$v cannot be sorted"))
+    size == IsInfinite() && throw(ArgumentError("infinite iterator $v cannot be sorted"))
+    sort!(copymutable(v); kws...)
+end
+sort(v::AbstractVector; kws...) = sort!(copymutable(v); kws...) # for method disambiguation
+sort(::AbstractString; kws...) =
+    throw(ArgumentError("sort(::AbstractString) is not supported"))
+sort(::Tuple; kws...) =
+    throw(ArgumentError("sort(::Tuple) is only supported for NTuples"))
+
+function sort(x::NTuple{N}; lt::Function=isless, by::Function=identity,
+              rev::Union{Bool,Nothing}=nothing, order::Ordering=Forward) where N
+    o = ord(lt,by,rev,order)
+    if N > 20
+        v = sort!(copymutable(x), DEFAULT_STABLE, o)
+        tuple((v[i] for i in 1:N)...)
+    else
+        _sort(x, o)
+    end
+end
```

Comment:
There seems no reason to limit this to `NTuple`. I think the `N > 20` branch should be used for non-homogeneous tuples as it's impossible to make it stable.

BTW, since the `merge` below is limited to `NTuple`, we'd better use generated function or something similar to reduce the compile overhead.
```julia
_sort(::Tuple{}, o::Ordering) = ()
_sort(x::Tuple{Any}, o::Ordering) = x
_sort(x::Tuple{Any,Any}, o::Ordering) = lt(o, x[2], x[1]) ? reverse(x) : x 
function _sort(x::Tuple, o::Ordering)
    x1, x2 = Base.IteratorsMD.split(x, Val(length(x)÷2))
    merge(_sort(x1, o), _sort(x2, o), o, Val(length(x)))
end
function merge(x::Tuple, y::Tuple, o::Ordering, L)
    i = Ref(1)
    j = Ref(1)
    return ntuple(L) do k
        if i[] > length(x) || (j[] <= length(y) && lt(o, y[j[]], x[i[]]))
            z = y[j[]]
            j[] += 1
        else
            z = x[i[]]
            i[] += 1
        end
        return z
    end
end
```

Reply:
> There seems no reason to limit this to `NTuple`.

Can you think of a case where folks would want to sort a non-homogenous tuple and get a non-homogenous tuple back>

Supporting this now would also be incompatible with making `sort((2.0,1))` automatically promote, be type stable, and return `(1.0,2.0)` in the future.

Reply:
Promotion is performative in this case, but it's strange to me that `sort` do promotion by default.